### PR TITLE
Basic Django Rest API

### PR DESCRIPTION
**Pull request by saadyousafarbi**
Basic Django API developed for model ```Profile``` with following functionality:

* API will be able to retrieve all Profiles within database for autoshop site.
* API will also have capability of editing profile with valid data provided.
* API will be only for authenticated users.

### CODE REVIEW

File: autoshop/api/v1/views.py [hunk 1]
```diff
@@ -0,0 +1,12 @@
+from rest_framework.views import APIView
+from rest_framework.response import Response
+from core.models import Profile
+from autoshop.api.v1.serializers import ProfileSerializer
```
Comment (zubair-arbi) [inline autoshop/api/v1/views.py]: ordering of imports

File: autoshop/api/v1/serializers.py [hunk 2]
```diff
@@ -0,0 +1,9 @@
+from core.models import Profile
+from rest_framework import serializers
+
+
+class ProfileSerializer(serializers.ModelSerializer):
```
Comment (zubair-arbi) [inline autoshop/api/v1/serializers.py]: We might want to rename it to `UserProfileSerializer` for clarity of whose profile its is.

File: autoshop/settings.py [hunk 3]
```diff
@@ -42,6 +42,8 @@
 
     'core.apps.LogistrationConfig',
     'analytics.apps.AnalyticsConfig',
+
+    'rest_framework',
```
Comment (zubair-arbi) [inline autoshop/settings.py]: the third party apps should come before local apps

File: autoshop/urls.py [hunk 4]
```diff
@@ -13,13 +13,19 @@
     1. Import the include() function: from django.conf.urls import url, include
     2. Add a URL to urlpatterns:  url(r'^blog/', include('blog.urls'))
 """
+from api.v1 import views
```
Comment (zubair-arbi) [inline autoshop/urls.py]: local imports at the end

File: autoshop/urls.py [hunk 5]
```diff
@@ -13,13 +13,19 @@
...
 
 
 urlpatterns = [
     url(r'^admin/', admin.site.urls),
     url(r'^', include('core.urls')),
+    url(r'^profiles/', views.ProfileList.as_view()),
 ] + static(settings.STATIC_URL, document_root=settings.STATIC_ROOT)
+
+
+urlpatterns = format_suffix_patterns(urlpatterns)
```
Comment (zubair-arbi) [inline autoshop/urls.py]: What is the purpose of `format_suffix_patterns`?

**Review by zubair-arbi — CHANGES_REQUESTED** [review]

**Comment by saadyousafarbi** [comment]
I have made corrections and also shifted from APIView to using generic views from rest framework. Please have a look at changes.

File: autoshop/api/v1/serializers.py [hunk 6]
```diff
@@ -0,0 +1,10 @@
+from rest_framework import serializers
+
+from core.models import Profile
+
+
+class UserProfileSerializer(serializers.ModelSerializer):
+
+    class Meta:
+        model = Profile
+        fields = '__all__'
```
Comment (zubair-arbi) [inline autoshop/api/v1/serializers.py:10]: I think we dont need time stamp model fields here. Also user should not be able to edit username which we use for login.

File: autoshop/api/v1/views.py [hunk 7]
```diff
@@ -0,0 +1,24 @@
+from rest_framework.generics import ListAPIView, RetrieveAPIView, UpdateAPIView, DestroyAPIView
+
+from autoshop.api.v1.serializers import UserProfileSerializer
+from core.models import Profile
+
+
+class ProfileList(ListAPIView):
```
Comment (zubair-arbi) [inline autoshop/api/v1/views.py]: I think the `ModelViewSet` is more appropriate for our use case.

Comment (zubair-arbi) [reply autoshop/api/v1/views.py]: Please look into these resources:
* http://www.django-rest-framework.org/api-guide/routers/#routers
* http://www.django-rest-framework.org/api-guide/viewsets/#modelviewset

**Review by zubair-arbi — CHANGES_REQUESTED** [review]

File: autoshop/api/v1/serializers.py (hunk 6 above)
Comment (saadyousafarbi) [reply autoshop/api/v1/serializers.py:10]: All important data is not present in Profile model. Username, email, password etc are present in Django's default User model and Profile model only has User ID associated with it other than profile information.

File: autoshop/api/v1/views.py (hunk 7 above)
Comment (saadyousafarbi) [reply autoshop/api/v1/views.py]: How do we determine which approach is better than the other? They all seem to achieve the desired output.

File: autoshop/api/v1/serializers.py (hunk 6 above)
Comment (zubair-arbi) [reply autoshop/api/v1/serializers.py:10]: But user should be able to edit name (`first_name` and `last_name`).

File: autoshop/api/v1/views.py (hunk 7 above)
Comment (zubair-arbi) [reply autoshop/api/v1/views.py]: Well its about what you want to do with end point. We want an endpoint for django model with the basic crud operations. Although we will not support the delete operation for now.

File: autoshop/api/v1/views.py [hunk 8]
```diff
@@ -0,0 +1,10 @@
+from rest_framework.viewsets import ModelViewSet
+
+from autoshop.api.v1.serializers import UserProfileSerializer
+from core.models import Profile
+
+
+class UserProfileViewSet(ModelViewSet):
```
Comment (zubair-arbi) [inline autoshop/api/v1/views.py:9]: implement actions provided by the `ModelViewSet`

File: autoshop/api/v1/serializers.py (hunk 6 above)
Comment (zubair-arbi) [reply autoshop/api/v1/serializers.py:10]: we still need to figure out how to include fields `first_name` and `last_name` with this API

File: autoshop/api/v1/views.py (hunk 8 above)
Comment (zubair-arbi) [reply autoshop/api/v1/views.py:9]: I am not seeing action for update

File: autoshop/api/v1/serializers.py (hunk 6 above)
Comment (zubair-arbi) [reply autoshop/api/v1/serializers.py:10]: Any follow up on this suggestion?

File: autoshop/api/v1/views.py [hunk 9]
```diff
@@ -0,0 +1,28 @@
...
+    queryset = Profile.objects.all()
+    serializer_class = UserProfileSerializer
+
+    def list(self, request):
+        queryset = Profile.objects.all()
```
Comment (zubair-arbi) [inline autoshop/api/v1/views.py:14]: do we need this `queryset`  same as default `queryset`?

Comment (saadyousafarbi) [reply autoshop/api/v1/views.py:14]: yes, it is required in every action.

File: autoshop/api/v1/serializers.py (hunk 6 above)
Comment (saadyousafarbi) [reply autoshop/api/v1/serializers.py:10]: We would have to change the model

**Review by zubair-arbi — APPROVED** [review]
Rebase, squash and merge 👍

**Comment by saadyousafarbi** [comment]
Rebased from 40d83256ecd8f4fe400aa75c135458661b860e8b

**Pull request merged**